### ISSUE TITLE
DNS queries leaks if always-on VPN with killswitch enabled malfunctions

### ISSUE THREAD

**Issue by ryrona2** [GrapheneOS/os-issue-tracker#3442]
I was testing out the VPN functionality while monitoring all network traffic on my Wifi hotspot, and while intentionally trying to sabotage the connection, just to see that the killswitch works as intended.

It turns out it is possible to cause DNS queries to leak outside the VPN when it malfunctions. This breaks the expected security a VPN with killswitch should provide, that your internet activity like what sites you visit remains hidden from your ISP even when VPN malfunctions.

Steps to reproduce:
1. Install Mullvad VPN app.
2. Login to Mullvad account and connect.
3. Use any app, like Vanadium web browser. No DNS requests leaks outside VPN.
4. Go to Settings, App permissions and remove Network permission for Mullvad VPN.
5. Use any app, like Vanadium web browser. Now DNS requests goes out outside VPN, even if no other network traffic works.

This cannot possibly be an issue with the Mullvad VPN app, since I have revoked its network permissions so it cannot be the one sending the DNS queries, so the leak is somewhere in GrapheneOS.

**Comment by ryrona2** [comment]
> Try using the official WireGuard app instead.

Unfortunately this didn't tell me anything, because the official Wireguard app detects when Network permission is removed from it, and immediately disables the VPN. No DNS is leaked in that case, but all apps believe there is no internet, so maybe they are not even trying. In the case of the Mullvad VPN app, it still believes there is internet, so doesn't disable its VPN configuration.

I would have expected the killswitch to block the DNS query anyway, however Mullvad VPN app has set up their configuration, since the killswitch should be blocking all traffic not going out over the VPN, with the VPN app having no say on the matter.

**Comment by Tryptamine9** [comment]
So you have confirmed that a correctly written VPN app, such as the official WireGuard app, that is known to function correctly DOES kill all network traffic when it looses network connection and the kill switch is activated! 

Also you have found that a poorly written VPN app DOES NOT kill all traffic when it looses connectivity, and there is a leak when DNS requests are made. Sounds to me like this should be filed with the development team behind Mulvad VPN, not here...

**Comment by ryrona2** [comment]
Sure, the Mullvad VPN app could probably act better here. But isn't the whole idea with a kill switch that apps shouldn't be able to make connections outside the VPN when the VPN is malfunctioning? The kill switch is named "Block connections without VPN". This is not happening here. DNS queries goes out without the VPN. And the kill switch is provided by GrapheneOS, not any VPN app. So even if this specific VPN app certainly could improve, it sounds like the actual security issue is in GrapheneOS, either by implementation or expectation of functionality.

**Comment by ryrona2** [comment]
> GrapheneOS uses the standard implementation of this with no changes to it. None of the features or bugs with VPN support have to do with GrapheneOS at the moment.

Okay. Do you expect me to file a bug ticket upstream, or do the GrapheneOS team handle it? Which bug tracker? I am not familiar with upstream here.

**Comment by Tryptamine9** [comment]
Kill switch is provided by AOSP, upstream of GrapheneOS. There are a few things that do go around the VPN kill switch, by design. When connected to a VPN, the VPN usually provides the DNS. When you are disconnected from it the VPN should work with the OS to ensure that there is no more network traffic going out, since the VPN cannot take it, it would leak.

The OpenVPN app I was trailing had an option that allowed pausing the VPN when screen was off. It warned that if the "persistent run" feature was available, then there would be no connectivity when VPN was off. If not, then data would leak out when VPN was off. Sounds like this is an example of the later...

Important to be not just using any old VPN app!

If you want to converse about this, please let's move to the discussion forum. The issue tracker is not for that.

**Comment by ryrona2** [comment]
@thestinger Yeah, I will investigate more before escalating to AOSP or app developer. The short investigation I did a few days ago showed that the VPN app detected the Network permission being removed, and tore down the VPN interface and then immediately set it up again. But the kill switch was not torn down, and it looks like it is set up in a proper way. And all the rest was set up in an identical way to before removing the Network permission, so I do not yet understand what caused DNS queries to suddenly go out outside VPN.

My current suspicion is that apps do not actually send DNS queries themselves, but that there is a system wide DNS resolver running as another UID which does all DNS queries, and something happens so apps suddenly are able to query the system wide DNS resolver, and that one was not bridged back to the VPN app properly during the rapid tear down and setup of the VPN interface, and since that one is not blocked by the kill switch, it ends up sending out the queries outside the VPN.

If you know how DNS is handled in GrapheneOS, please tell me, since that could speed up my investigations a little. Also if you know how the removal of Network permissions is implemented. I have at least confirmed the permission removal isn't implemented the same way as the kill switch, but still don't know how. I initially suspected they may clash, but that is probably not what is happening after all.

I will prioritize the multicast leak ticket since that is a confirmed real problem. For this one, if the issue is only happening if Network permission is removed from the VPN app, it is not such a serious issue, since there is no logical reason a user would remove the Network permission for the VPN app. Still a bug, but not so serious. I will try to find another way to reproduce the issue that works on AOSP too, but I guess the easiest way is to first find what the actual issue is.

**Comment by Rawa** [comment]
Hello! I'm a developer from Mullvad VPN whom been looking into this issue over the last week and thought I might shim in and give some more context. 

The DNS leak is possible to reproduce in the Wireguard app as well. We have reported the issue to Google, you can find the issue here, including steps to reproduce:
https://issuetracker.google.com/issues/337961996

From our testing we can observe the following (also stated in our issue to Google):
1. If no DNS server is configured on the VPN, DNS requests may leak.
2. When a tunnel is torn down, the system will leak as well. So setting up two tunnels in the Wireguard app, both with DNS and then switching between them will cause a DNS leak as well.

From our testing we see that the android [DNSResolver](https://developer.android.com/reference/android/net/DnsResolver) and [DatagramSocket](https://developer.android.com/reference/java/net/DatagramSocket) with [DatagramPacket](https://developer.android.com/reference/java/net/DatagramPacket) won't go out, but some browsers will leak (e.g Chrome), maybe because it uses the Native DNS as mentioned by @thestinger.

Also below you can find a gist with a HTML file that has javascript embedded. The HTML file can be opened in your browser of choice and will do GET requests to unique URLs, thus resulting in new DNS requests. Running this in Chrome with case 1 & 2 and observing network traffic with e.g `tcpdump` you can see this leak in action.
https://gist.github.com/Rawa/dcc636e45f95143a8ea65ba3ca366ae8

Thanks for creating this issue request and reporting it also to us.

**Comment by ryrona2** [comment]
> It's entirely supported to send DNS queries to the regular network DNS while using a VPN with the kill switch enabled.

I just want to say this is totally unexpected behavior from the user's perspective, whether it is as designed or not.

If you set up a kill switch for the VPN on Linux, you would add an iptables rule that blocks all network traffic not going to the specific IP:port combination belonging to the VPN. This would bring peace in mind that even if something does go wrong with the VPN, the system DNS resolver nor anything else can send out traffic on the network.

Also, the kill switch is really just a poor approximation of how a VPN setup should be really designed. Whonix is doing it right with their Workstation and Gateway VMs, and QubesOS supports doing a similar setup for VPNs. That is leak proof in every single way.

**Comment by mateusz-markowicz** [comment]
We did some testing on our side (ProtonVPN) and what's happening is that after establishing the tunnel all API requests that our app does fail with DNS errors. When openining the tunnel we set in VpnService.Builder DNS server to 10.2.0.1 (see https://developer.android.com/reference/android/net/VpnService.Builder#addDnsServer(java.net.InetAddress)). We noticed that when the app doesn't set DNS server at all or we set it to some public DNS like 1.1.1.1 everything works fine. So it seems like your fix somehow interferes with the way we set DNS server for the connection.

**Comment by albin-mullvad** [comment]
Hey! I'm working with @Rawa on the Mullvad VPN Android app. We also did some testing and could see similar results as reported by @mateusz-markowicz. By the time we were going to report back here a few days ago we saw that the related fix already had been reverted (https://github.com/GrapheneOS/platform_system_netd/commit/296ccdc5eb955f67ff8ced0a2612e90a5bd77624).

While the fix was addressing the original issue, we could see that regular API communication was broken while having a connected tunnel. In our test setup we could see that DNS queries would leak outside the tunnel BUT still use the tunnel DNS server. In the following capture, `192.168.1.128` is the device and `10.64.0.1` is the DNS server configured for the tunnel.

`
09:32:59.797907 IP 192.168.1.128.10157 > 10.64.0.1.53: 42899+ A? ipv4.am.i.mullvad.net. (39)
`

**Comment by EnfLzSp** [comment]
About Multicast:
Why not just add local network isolation toggle to block all multicast traffic system-wide or on per-app basis and leave Network Permission function as it is? Multicast shouldn't be an issue on mobile networks as everyone is supposed to be isolated from each other, but WiFi AP's are a different matter. An option to disable WiFi Direct would also be nice.

About DNS Leakd:
DNS queries outside of VPN tunnels are leaks, even if they are made to VPN DNS, but it looks like some VPN apps don't agree with that.
